drivers/display/ssd1327.c

```diff
@@ -67,186 +80,256 @@ static inline int ssd1327_write_bus_cmd(const struct device *dev, const uint8_t
-		return -EIO;
+	err = config->write_cmd(dev, SSD1327_SET_DISPLAY_OFFSET, &config->display_offset, 1);
+	if (err < 0) {
+		return err;
 	}
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_NORMAL_DISPLAY, NULL, 0)) {
-		return -EIO;
+	err = config->write_cmd(dev, SSD1327_SET_NORMAL_DISPLAY, NULL, 0);
+	if (err < 0) {
+		return err;
 	}
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_SEGMENT_MAP_REMAPED, &config->remap_value, 1)) {
-		return -EIO;
+	err = config->write_cmd(dev, SSD1327_SET_SEGMENT_MAP_REMAPED, &config->remap_value, 1);
+	if (err < 0) {
+		return err;
 	}
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_MULTIPLEX_RATIO, &config->multiplex_ratio, 1)) {
-		return -EIO;
+	err = config->write_cmd(dev, SSD1327_SET_MULTIPLEX_RATIO, &config->multiplex_ratio, 1);
+	if (err < 0) {
+		return err;
 	}
 	buf = SSD1327_ENABLE_VDD;
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_FUNCTION_A, &buf, 1)) {
-		return -EIO;
-	}
-
-	return 0;
+	return config->write_cmd(dev, SSD1327_SET_FUNCTION_A, &buf, 1);
 }
 
 static int ssd1327_resume(const struct device *dev)
 {
-	return ssd1327_write_bus_cmd(dev, SSD1327_DISPLAY_ON, NULL, 0);
+	const struct ssd1327_config *config = dev->config;
+
+	return config->write_cmd(dev, SSD1327_DISPLAY_ON, NULL, 0);
 }
 
 static int ssd1327_suspend(const struct device *dev)
 {
-	return ssd1327_write_bus_cmd(dev, SSD1327_DISPLAY_OFF, NULL, 0);
+	const struct ssd1327_config *config = dev->config;
+
+	return config->write_cmd(dev, SSD1327_DISPLAY_OFF, NULL, 0);
 }
 
 static int ssd1327_set_display(const struct device *dev)
 {
 	const struct ssd1327_config *config = dev->config;
-	uint8_t x_position[] = {
-		0,
-		config->width - 1
-	};
-	uint8_t y_position[] = {
-		0,
-		config->height - 1
-	};
-
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_COLUMN_ADDR, x_position, sizeof(x_position))) {
-		return -EIO;
+	int err;
+	uint8_t x_position[] = {0, config->width - 1};
+	uint8_t y_position[] = {0, config->height - 1};
+
+	err = config->write_cmd(dev, SSD1327_SET_COLUMN_ADDR, x_position, sizeof(x_position));
+	if (err < 0) {
+		return err;
 	}
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_ROW_ADDR, y_position, sizeof(y_position))) {
-		return -EIO;
+	err = config->write_cmd(dev, SSD1327_SET_ROW_ADDR, y_position, sizeof(y_position));
+	if (err < 0) {
+		return err;
 	}
-	if (ssd1327_write_bus_cmd(dev, SSD1327_SET_SEGMENT_MAP_REMAPED, &config->remap_value, 1)) {
-		return -EIO;
+	return config->write_cmd(dev, SSD1327_SET_SEGMENT_MAP_REMAPED, &config->remap_value, 1);
+}
+
+#if DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(solomon_ssd1327fb, mipi - dbi)
+static int ssd1327_write_pixels_mipi(const struct device *dev, const uint8_t *buf,
+				     uint32_t pixel_count,
+				     const struct display_buffer_descriptor *desc)
+{
+	const struct ssd1327_config *config = dev->config;
+	struct display_buffer_descriptor mipi_desc;
+	int ret;
+	int total = 0;
+
+	mipi_desc.pitch = desc->pitch / 2;
+
+	while (pixel_count > total) {
```

Comment:
This is non trivial to understand what you are doing, a comment would be good. Also you are doing the same stuff further down again, maybe extract to method?

Reply:
I dont even know if the SPI code works even for now :/
In ssd1322 this is a another function, i referenced it but didnt like how it was organized, as it takes a couple pointers to variables from the loop context, I will try thinking of a better organization.

Reply:
Well i confirmed this exact code works on SSD1320 (which i have in SPI flavor and used the exact same code for copy to device with), but i still havent figured a better organization

Reply:
Is this new way satisfactory?

Reply:
@faxe1008 ?